url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/fastme/tree.java

```diff
@@ -0,0 +1,1616 @@
+		   		}
+		   		if (null != right)
+		   		{
+		   			updateSubTreeAverages(A, right, v, direction.UP);
+		   			A[par.head.index][right.head.index]
+		   			    = A[right.head.index][par.head.index]
+		   			    = (par.topsize*A[par.head.index][right.head.index]
+		   			    + A[right.head.index][v.index]) / (par.topsize + 1);
+		   		}
+		   		break;
+		   	case SKEW: //point of insertion is skew to e
+			    // 1-distant average of nodes below e to 
+				// nodes above e
+		   		A[e.head.index][e.head.index] = 
+		   			(e.topsize*A[e.head.index][e.head.index] + 
+		   			A[e.head.index][v.index])/(e.topsize + 1);      
+			    //no 2-distant averages to update in this case
+			    //updating 3-distant averages involving sib
+		   		if (null != left)
+		   		{
+		   			updateSubTreeAverages(A, left, v, direction.UP);
+		   			A[sib.head.index][left.head.index]
+		   			    = A[left.head.index][sib.head.index]
+	                    = (sib.bottomsize*A[sib.head.index][left.head.index]
+	                    + A[left.head.index][v.index]) / (sib.bottomsize + 1);
+		   		}
+		   		if (null != right)
+		   		{
+		   			updateSubTreeAverages(A, right, v, direction.UP);
+		   			A[sib.head.index][right.head.index]
+		   			    = A[right.head.index][sib.head.index]
+		   			    = (sib.bottomsize*A[par.head.index][right.head.index]
+		   			    + A[right.head.index][v.index]) / (sib.bottomsize + 1);
+		   		}
+		   		break;
+		   	case LEFT: //point of insertion is below the edge left
+		   		//1-distant average
+		   		A[e.head.index][e.head.index] = 
+		   			(e.bottomsize*A[e.head.index][e.head.index] + 
+		   			A[v.index][e.head.index])/(e.bottomsize + 1);        
+		   		// 2-distant averages
+		   		A[e.head.index][e.tail.index] = 
+		   			A[e.tail.index][e.head.index] = 
+		   			(e.bottomsize*A[e.head.index][e.tail.index] + 
+		   			A[v.index][e.tail.index])/(e.bottomsize + 1);  
+		   		A[left.head.index][right.head.index] = 
+		   			A[right.head.index][left.head.index] = 
+		   			(left.bottomsize*A[right.head.index][left.head.index]
+		   			+ A[right.head.index][v.index]) / (left.bottomsize+1);
+		   		//3-distant avereages involving left
+		   		if (null != sib)
+		   		{
+		   			updateSubTreeAverages(A, sib, v, direction.SKEW);
+		   			A[left.head.index][sib.head.index]
+		   			     = A[sib.head.index][left.head.index]
+		   			     = (left.bottomsize*A[left.head.index][sib.head.index]
+		   			     + A[sib.head.index][v.index]) / (left.bottomsize + 1);
+		   		}
+		   		if (null != par)
+		   		{
+		   			if (e.tail.leftEdge == e)
+		   				updateSubTreeAverages(A, par, v, direction.LEFT);
+		   			else
+		   				updateSubTreeAverages(A, par, v, direction.RIGHT);
+		   			A[left.head.index][par.head.index]
+		   			     = A[par.head.index][left.head.index]
+		   			     = (left.bottomsize*A[left.head.index][par.head.index]
+		   			     + A[v.index][par.head.index]) / (left.bottomsize + 1);
+		   		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 419:429, 454:464.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.